libmamba/src/core/satisfiability_error.cpp

```diff
@@ -0,0 +1,331 @@
+                    + (m_build_range.empty() ? 0 : 1) + m_version_range.size());
+        if (!m_version_range.empty())
+        {
+            out += ' ';
+            out += m_version_range;
+        }
+        if (!m_build_range.empty())
+        {
+            out += ' ';
+            out += m_build_range;
+        }
+        return out;
+    }
+
+    namespace
+    {
+
+        auto warn_unexpected_problem(MSolverProblem const& problem) -> void
+        {
+            // TODO: Once the new error message are not experimental, we should consider
+            // reducing this level since it is not somethig the user has control over.
+            LOG_WARNING << "Unexpected empty optionals for problem type "
+                        << solver_ruleinfo_name(problem.type);
+        }
+
+        class ProblemsGraphCreator
+        {
+        public:
+            using SolvId = Id;  // Unscoped from libsolv
+
+            using graph_t = ProblemsGraph::graph_t;
+            using RootNode = ProblemsGraph::RootNode;
+            using PackageNode = ProblemsGraph::PackageNode;
+            using UnresolvedDependencyNode = ProblemsGraph::UnresolvedDependencyNode;
+            using ConstraintNode = ProblemsGraph::ConstraintNode;
+            using node_t = ProblemsGraph::node_t;
+            using node_id = ProblemsGraph::node_id;
+            using edge_t = ProblemsGraph::edge_t;
+            using conflict_map = ProblemsGraph::conflict_map;
+
+            ProblemsGraphCreator(MSolver const& solver, MPool const& pool)
+                : m_solver{ solver }
+                , m_pool{ pool }
+            {
+                m_root_node = m_graph.add_node(RootNode());
+                parse_problems();
+            }
+
+            operator ProblemsGraph() &&
+            {
+                return { std::move(m_graph), std::move(m_conflicts), m_root_node };
+            }
+
+        private:
+            MSolver const& m_solver;
+            MPool const& m_pool;
+            graph_t m_graph;
+            conflict_map m_conflicts;
+            std::map<SolvId, node_id> m_solv2node;
+            node_id m_root_node;
+
+            /**
+             * Add a node and return the node id.
+             *
+             * If the node is already present and ``update`` is false then the current
+             * node is left as it is, otherwise the new value is inserted.
+             */
+            auto add_solvable(SolvId solv_id, node_t&& pkg_info, bool update = true) -> node_id;
+
+            auto add_conflict(node_id n1, node_id n2) -> void;
+            [[nodiscard]] auto add_expanded_deps_edges(node_id from_id,
+                                                       SolvId dep_id,
+                                                       edge_t const& edge) -> bool;
+
+            auto parse_problems() -> void;
+        };
+
+        auto ProblemsGraphCreator::add_solvable(SolvId solv_id, node_t&& node, bool update)
+            -> node_id
```

Comment:
Same here

Reply:
This one is a dependent type, would you prefer the following?
```cpp
ProblemsGraphCreato::node_id ProblemsGraphCreator::add_solvable(SolvId solv_id, node_t&& node, bool update);
```

Reply:
Ha my bad, in that case the trailing return type is better.